tests/integration/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php

```diff
@@ -0,0 +1,425 @@
+            $this->getContainer()->get(ProductDefinition::class),
+            $criteria,
+            self::createLocalizedContext([
+                $this->getDeDeLanguageId(),
+                Defaults::LANGUAGE_SYSTEM,
+            ]),
+        );
+
+        static::assertEquals([$productId1], $productIds->getIds());
+    }
+
+    private function createProduct(
+        string $productNumber,
+        ?string $deDeTranslation,
+        ?string $defaultTranslation = null,
+        array $categories = [],
+        ?string $parentProductNumber = null,
+        ?IdsCollection $ids = null,
+    ): string {
+        $ids ??= new IdsCollection();
+        $productBuilder = $this->buildProduct(
+            deDeTranslation: $deDeTranslation,
+            defaultTranslation: $defaultTranslation,
+            productNumber: $productNumber,
+            parentProductNumber: $parentProductNumber,
+            ids: $ids,
+        );
+        $productBuilder->categories($categories);
+
+        $this->getContainer()->get('product.repository')->create(
+            [$productBuilder->build()],
+            Context::createDefaultContext()
+        );
+
+        return $ids->get($productNumber);
+    }
+
+    private function buildProduct(
+        ?string $deDeTranslation = null,
+        ?string $defaultTranslation = null,
+        string $productNumber = 'product-1',
+        ?string $parentProductNumber = null,
+        ?IdsCollection $ids = null,
+    ): ProductBuilder {
+        $ids ??= new IdsCollection();
+        $productBuilder = new ProductBuilder($ids, $productNumber);
+        $productBuilder->price(100);
+        if ($deDeTranslation !== null) {
+            $productBuilder->translation($this->getDeDeLanguageId(), 'name', $deDeTranslation);
+        }
+        if ($defaultTranslation !== null) {
+            $productBuilder->translation(Defaults::LANGUAGE_SYSTEM, 'name', $defaultTranslation);
+        }
+        if ($parentProductNumber !== null) {
+            $productBuilder->parent($parentProductNumber);
+        }
+
+        return $productBuilder;
+    }
+
+    private function createCategory(
+        string $defaultTranslation,
+        ?string $deDeTranslation,
+        ?IdsCollection $ids = null
+    ): string {
+        $ids ??= new IdsCollection();
+        // Category does not have a name filed but only translations, hence the default translation must be passed to
+        // the builder as the name
+        $categoryBuilder = new CategoryBuilder($ids, categoryName: $defaultTranslation);
+        if ($deDeTranslation !== null) {
+            $categoryBuilder->translation($this->getDeDeLanguageId(), 'name', $deDeTranslation);
+        }
+
+        $this->getContainer()->get('category.repository')->create(
+            [$categoryBuilder->build()],
+            Context::createDefaultContext(),
+        );
+
+        return $ids->get($defaultTranslation);
+    }
+
+    /**
+     * @param string[] $languageIdChain
```

Comment:
```suggestion
     * @param list<string> $languageIdChain
```

this annotation make it more clearer how the array looks like. in PHPStans view `string[]` would also be valid for `array<string, string>` types